File: 108820008_HW12/108820008_HW12-2/108820008_HW12-2/app/src/main/java/com/example/downloadscheduler/MainActivity.java
package com.example.downloadscheduler;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NotificationCompat;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.job.JobInfo;
import android.app.job.JobParameters;
import android.app.job.JobScheduler;
import android.app.job.JobService;
import android.content.ComponentName;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {
    public static final int JOB_ID = 0;
    private static final String PRIMARY_CHANNEL_ID = "primary_notification_channel";
    NotificationManager manager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        createNotificationChannel();

        final Button downloadButton = findViewById(R.id.button_download);
        downloadButton.setOnClickListener(view -> {
            sendDownloadNotification();
            scheduleDownloadJob();
        });
    }

    private void createNotificationChannel() {
        manager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel notificationChannel = new NotificationChannel(PRIMARY_CHANNEL_ID, getResources().getString(R.string.channel_name), NotificationManager.IMPORTANCE_HIGH);
            notificationChannel.enableLights(true);
            notificationChannel.setLightColor(Color.RED);
            notificationChannel.enableVibration(true);
            notificationChannel.setDescription("Downloading task running notification");
            manager.createNotificationChannel(notificationChannel);
        }
    }

    public void scheduleDownloadJob() {
        JobScheduler scheduler = (JobScheduler) getSystemService(JOB_SCHEDULER_SERVICE);
        ComponentName componentName = new ComponentName(getApplicationContext(), DownloadJobService.class.getName());
        JobInfo.Builder builder = new JobInfo.Builder(JOB_ID, componentName);
        builder.setRequiresCharging(true)
                .setRequiredNetworkType(JobInfo.NETWORK_TYPE_UNMETERED)
                .setRequiresDeviceIdle(true)
                .setPeriodic(TimeUnit.DAYS.toMillis(1));
        if (Build.VERSION.SDK_INT > 23) {
            builder.setPeriodic(TimeUnit.DAYS.toMillis(1), TimeUnit.MINUTES.toMillis(5));
        }
        JobInfo jobInfo = builder.build();
        if (scheduler != null) {
            scheduler.schedule(jobInfo);
        }
    }

    public void sendDownloadNotification() {
        Intent intent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, PRIMARY_CHANNEL_ID)
                .setContentTitle("Performing Work")
                .setContentText("Downloading in progress")
                .setContentIntent(pendingIntent)
                .setSmallIcon(R.drawable.ic_adb)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setDefaults(NotificationCompat.DEFAULT_ALL)
                .setAutoCancel(true);
        manager.notify(0, builder.build());
    }

    public class DownloadJobService extends JobService {
        @Override
        public boolean onStartJob(JobParameters params) {
            sendDownloadNotification();
            return true;
        }

        @Override
        public boolean onStopJob(JobParameters params) {
            return false;
        }
    }
}

File: 108820008_HW8/108820008_HW8-1/108820008_HW8-1/app/src/main/java/com/example/battery/MainActivity.kt
package com.example.battery

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    private lateinit var batteryImageView: ImageView
    private lateinit var plusButton: Button
    private lateinit var minusButton: Button
    private var batteryLevel = 4
    private val minBatteryLevel = 0
    private val maxBatteryLevel = 7

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        batteryImageView = findViewById(R.id.image_battery)
        plusButton = findViewById(R.id.button_plus)
        minusButton = findViewById(R.id.button_minus)
        batteryImageView.setImageLevel(batteryLevel)

        plusButton.setOnClickListener {
            batteryLevel =
                if (batteryLevel + 1 > maxBatteryLevel) maxBatteryLevel else batteryLevel + 1
            batteryImageView.setImageLevel(batteryLevel)
        }

        minusButton.setOnClickListener {
            batteryLevel =
                if (batteryLevel - 1 < minBatteryLevel) minBatteryLevel else batteryLevel - 1
            batteryImageView.setImageLevel(batteryLevel)
        }
    }
}